casper/src/main/scala/coop/rchain/casper/finality/Finalizer.scala

```diff
@@ -0,0 +1,216 @@
+  * if it has more then half of total stake in bonds map of A translated from tips throughout main parent chain.
+  * IMPORTANT: only main parent relation gives weight to potentially finalized block.
+  *
+  * Therefore: Finalizer should seek for next LFB going through 2 steps:
+  *   1. Find messages in scope of search that have more then half of the stake translated through main parent chain
+  *     from tips down to the message.
+  *   2. Execute [[CliqueOracle.computeOutput]] on these targets.
+  *   3. Candidate with the highest block number passing FT becomes the next LFB.
+  */
+object Finalizer {
+
+  /** Message that is agreed on + weight of this agreement. */
+  final case class Agreement[M, V](message: M, stakeAgreed: WeightMap[V])
+
+  /** If more then half of total stake agree on message - it is considered to be safe from orphaning. */
+  def cannotBeOrphaned[F[_]: Applicative, M, V](
+      messageWeightMap: WeightMap[V],
+      agreeingWeightMap: WeightMap[V]
+  ): Boolean = {
+    require(
+      !agreeingWeightMap.exists { case (_, stake) => stake <= 0 },
+      message = "Agreeing map contains not bonded validators"
+    )
+    val activeStakeTotal    = messageWeightMap.values.sum
+    val activeStakeAgreeing = agreeingWeightMap.values.sum
+    // in theory if each stake is high enough, e.g. Long.MaxValue, sum of them might result in negative value
+    require(activeStakeTotal > 0, message = "Long overflow when computing total stake")
+    require(activeStakeAgreeing > 0, message = "Long overflow when computing total stake")
+    activeStakeAgreeing > activeStakeTotal.toFloat / 2
+  }
+
+  /**
+    * Create an agreement given validator that agrees on a message and weight map of a message.
+    * If validator is not present in message bonds map or its stake is zero, None is returned
+    */
+  def recordAgreement[V](
+      messageWeightMap: WeightMap[V],
+      agreeingValidator: V
+  ): Option[WeightMap[V]] = {
+    // if validator is not bonded according to message weight map - there is no agreement translated.
+    val stakeAgreed = messageWeightMap.getOrElse(agreeingValidator, 0L)
+    if (stakeAgreed > 0) Map(agreeingValidator -> stakeAgreed).some else None
+  }
+
+  /**
+    * Agreements propagated by message to its ancestors.
+    * @param message message that agrees
+    * @param validator creator of a message
+    * @param dag message dependency graph
+    * @param stopHeight height of message to stop traversing the DAG
+    * @param heightF function to get message height
+    * @param messageWeightMapF function to get bonds map (weights map) for some message
+    * @return Stream of agreements propagated by message onto its ancestors down till stopHeight
+    */
+  def messageAgreements[F[_]: Sync, M, V](
+      message: M,
+      validator: V,
+      dag: DagReader[F, M],
+      stopHeight: Long,
+      heightF: M => F[Long],
+      messageWeightMapF: M => F[WeightMap[V]]
+  ): Stream[F, Agreement[M, V]] =
```

Comment:
Output of this can be cached, its defined by message and stopHeight (which is LFB that should change in a discrete way)